src/mpyl/projects/versioning.py

```diff
@@ -0,0 +1,149 @@
+"""Versioning and upgrade utilities for mpyl projects.
+### Writing an upgrade script
+
+To write an upgrade script, create a new class that inherits from `Upgrader` and
+implements the `upgrade` method. This class should then be added to the `UPGRADERS`
+list in this module.
+
+"""
+import copy
+import pkgutil
+from abc import ABC
+from pathlib import Path
+from typing import Optional, Generator
+
+from deepdiff import DeepDiff
+from ruamel.yaml.compat import ordereddict
+
+from ..utilities.yaml import yaml_to_string, load_for_roundtrip
+
+VERSION_FIELD = "mpylVersion"
+
+
+def get_releases() -> list[str]:
+    embedded_releases = pkgutil.get_data(__name__, "releases/releases.txt")
+    if not embedded_releases:
+        raise ValueError("File releases/releases.txt not found in package")
+    releases = embedded_releases.decode("utf-8").strip().splitlines()
+    return list(reversed(releases))
+
+
+def get_latest_release() -> str:
+    return get_releases()[0]
+
+
+class Upgrader(ABC):
+    """Base class for upgrade scripts"""
+
+    target_version: str
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+class Upgrader11(Upgrader):
+    target_version = "1.0.11"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+class Upgrader10(Upgrader):
+    target_version = "1.0.10"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        if found_deployment := previous_dict.get("deployment", {}):
+            existing_namespace = found_deployment.get("namespace", None)
+            if not existing_namespace:
+                previous_dict["deployment"].insert(
+                    0, "namespace", previous_dict["name"]
+                )
+
+        return previous_dict
+
+
+class Upgrader9(Upgrader):
+    target_version = "1.0.9"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        previous_dict.insert(3, VERSION_FIELD, self.target_version)
+        return previous_dict
+
+
+class Upgrader8(Upgrader):
+    target_version = "1.0.8"
+
+    def upgrade(self, previous_dict: ordereddict) -> ordereddict:
+        return previous_dict
+
+
+UPGRADERS = [Upgrader8(), Upgrader9(), Upgrader10(), Upgrader11()]
+
+
+def get_entry_upgrader_index(
```

Comment:
why return the index instead of just the `Upgrader` you need?

Reply:
Because we need to loop over all upgraders in order, starting with the version in the `project.yml` or the base.